### ISSUE TITLE
RFC: How to handle "template" files whe installing open-iscsi?

### ISSUE THREAD

**Issue by gonzoleeman** [open-iscsi/open-iscsi#323]
I am working on adding the ability to easily move files currently installed in /etc/iscsi to other places, as passed in on the top-level "make" command. E.g. this would move the Database files to /var/lib/iscsi:

`sh> make DBDIR=/var/lib/iscsi all`
`sh> sudo make DBDIR=/var/lib/iscsi install`

My question is about what I call template files. These are files that (currently) have '@SBINDIR@' in them, where we have to translate that to a real directory ("/sbin" by default).

Right now, for example, the source file etc/system/iscsi-init.service has '@SBINDIR@' in it, which needs to be translated to /usr/lib/systemd/system/iscsi-init.service.

Should the template files be called something different, e.g. "iscsi-init.service_tmpl"? Does anyone care?

My problem is that etc/systemd/iscsi-init.service currently is really a template, and I don't want any confusion. Does anyone else care?

Any input would be most welcome, as I get too focused on this when doing the work in a vacuum, and no telling what I might come up with without outside feedback. :)

So, bottom line, should template files be names specially, or is the current approach fine?

TIA!

**Comment by gonzoleeman** [comment]
Is it time to take the plunge and use autoconf for this? :(

**Comment by gonzoleeman** [comment]
For the record, right now there is one thing I am making configurable from the "make" top level:

1. "sbindir" -- the "sbin" directory. Defaults to /sbin, but can be, for example, /usr/sbin

I would like to add to this list, two more things that could be configured:

1. "homedir" -- the "home" directory -- where iscsid.conf lives, as well as initiatorname.iscsi, and lock files
2. "dbdir" -- the "database" directory, where node, discovery, firmware, etc, databases live

Currently, by default, both are /etc/iscsi. It seems that RH uses /usr/lib/iscsi for the "dbdir".

Because there will be more things that need configuration, I brought up this issue.

**Comment by wenchao-hao** [comment]
> Right now, for example, the source file etc/system/iscsi-init.service has '@sbindir@' in it, which needs to be translated to /usr/lib/systemd/system/iscsi-init.service.
> 
> Should the template files be called something different, e.g. "iscsi-init.service_tmpl"? Does anyone care?

I am confused with iscsi-init.service. If it is a template file, then I think we should not add "Requires=iscsi-init.service" in iscsid.service. In my scenario, we do not need iscsi-init.service. But if we do not install it, following error would be reported, so I have to modify iscsid.service.

```
Failed to restart iscsi.service: Unit iscsi-init.service not found.
```

**Comment by wenchao-hao** [comment]
> For the record, right now there is one thing I am making configurable from the "make" top level:
> 
> 1. "sbindir" -- the "sbin" directory. Defaults to /sbin, but can be, for example, /usr/sbin
> 
> I would like to add to this list, two more things that could be configured:
> 
> 1. "homedir" -- the "home" directory -- where iscsid.conf lives, as well as initiatorname.iscsi, and lock files
> 2. "dbdir" -- the "database" directory, where node, discovery, firmware, etc, databases live
> 
> Currently, by default, both are /etc/iscsi. It seems that RH uses /usr/lib/iscsi for the "dbdir".
> 
> Because there will be more things that need configuration, I brought up this issue.

What about using environment variables such as ISCISI_DB, ISCSI_CONFIG, ISCSI_HOME? Actually, we once received a complaint from our users because the home path changed, so it means a lot to provide a way which do not affect users who are already using /etc/iscsi.

**Comment by gonzoleeman** [comment]
>> ... 
> I am confused with iscsi-init.service. If it is a template file, then I think we should not add "Requires=iscsi-init.service" in iscsid.service. In my scenario, we do not need iscsi-init.service. But if we do not install it, following error would be reported, so I have to modify iscsid.service to it run.
> 
> ```
> Failed to restart iscsi.service: Unit iscsi-init.service not found.
> ```

You are mixing up two different issues. The issue of whether or not iscsi-init.service is needed is separate.

Let's take iscsid.service as an example instead, and ignore iscsi-init.

The whole reason for these template files is that some use /sbin for the binary directory, and some use /usr/sbin. So how to handle this difference is the question. Managing something like this in man pages, shell scripts, and systemd service files is the issue.

Right now, what I did was to put "@SBINDIR@' in those places, and then use 'sed' to replace that string with either "/sbin" or "/usr/sbin" (or whatever is passed in) at "make install" time.

But we are about to get more things that need substituting at install time, so I want this to be as clean, concise, and easy to manage as possible.

One way to fix this is to continue doing what is being done, but with more files and more strings to substitute.

Another way is to do the same thing, but name the template files differently, so you know they are template files.

Lastly, it could be addressed with autoconf/automake/... That's probably the "right" way to do it, but that adds some complication and maintenance over the simpler solutions.

**Comment by gonzoleeman** [comment]
> > For the record, right now there is one thing I am making configurable from the "make" top level:
> > 
> > 1. "sbindir" -- the "sbin" directory. Defaults to /sbin, but can be, for example, /usr/sbin
> > 
> > I would like to add to this list, two more things that could be configured:
> > 
> > 1. "homedir" -- the "home" directory -- where iscsid.conf lives, as well as initiatorname.iscsi, and lock files
> > 2. "dbdir" -- the "database" directory, where node, discovery, firmware, etc, databases live
> > 
> > Currently, by default, both are /etc/iscsi. It seems that RH uses /usr/lib/iscsi for the "dbdir".
> > Because there will be more things that need configuration, I brought up this issue.
> 
> What about using environment variables such as ISCISI_DB, ISCSI_CONFIG, ISCSI_HOME? Actually, we once received a complaint from our users because the home path changed, so it means a lot to provide a way which do not affect users who are already using /etc/iscsi.

Again, this is two different issues. I want to add the ability to move these directories, but that does not require them to actually move, at least not yet. It's up to each distribution to decide where it wants files located. For example, right now RedHat seems to put the database files in /var/lib/iscsi, whereas the default is /etc/iscsi. Also, some distributions are coming up with new rules that make it harder to put stuff in the /etc directory (like openSUSE Tumbleweed). But yes, of course, users don't like change. In my case, In my case, I will likely first move the database files from /etc/iscsi to /var/lib/iscsi and leave the configuration and initiatorname files still in /etc/iscsi. I don't think users will care about that.

But you cannot change the location of the database at run time unless you really know what you're doing. In general, that would be like switching databases between uses.
Are you suggesting using environment variables at build time or at run time?

**Comment by gonzoleeman** [comment]
Okay, I've decided how I should do this, at least for now.

It does **NOT** involve autoconf, since autoconf solves a different set of issues having to do with how to compile things differently on different platforms. Instead, I'll continue to use "sed" to translate "template" files to real files, but I'll clean it up a little.

The new approach will be
1. rename the files being used as "templates" with a suffix of "template"
2. running "make" will generate the local file from the template (using "sed")
3. running "make install" will install the generated file
4. I will use patterns in the Makefiles to accomplish the transformations

I will clean up the Makefile in util and add a Makefile in etc, and move most of the subdirectory-specific stuff in the top-level Makefile to sub-Makefiles where the work is needed.

There is currently only one thing to be translated from the templates, and thats the "sbin" directory. But soon I'll add "homedir" and "dbdir" (both defaulting to "/etc/iscsi").

I am closing this issue.
